File: sami/step_definitions/Multiple Payment Methods.js
Given('the user is on the checkout page', () => {
  cy.visit('/checkout');
});

When('the user selects {string} as the payment method', (method) => {
  cy.get('#payment-method').select(method);
});

And('the user enters valid credit card details', () => {
  cy.get('#card-number').type('4111111111111111');
  cy.get('#expiry-date').type('12/25');
  cy.get('#cvv').type('123');
});

And('the user enters valid debit card details', () => {
  cy.get('#card-number').type('5500000000000004');
  cy.get('#expiry-date').type('12/25');
  cy.get('#cvv').type('123');
});

And('the user enters invalid credit card details', () => {
  cy.get('#card-number').type('1234567890123456');
  cy.get('#expiry-date').type('12/25');
  cy.get('#cvv').type('123');
});

And('the user enters invalid debit card details', () => {
  cy.get('#card-number').type('1234567890123456');
  cy.get('#expiry-date').type('12/25');
  cy.get('#cvv').type('123');
});

And('the user submits the payment', () => {
  cy.get('#submit-payment').click();
});

And('the user is redirected to the PayPal login page', () => {
  cy.url().should('include', 'paypal.com');
});

And('the user logs in with valid PayPal credentials', () => {
  cy.get('#paypal-email').type('user@example.com');
  cy.get('#paypal-password').type('password');
  cy.get('#paypal-login').click();
});

And('the user confirms the payment', () => {
  cy.get('#confirm-payment').click();
});

And('the user cancels the payment', () => {
  cy.get('#cancel-payment').click();
});

Then('the payment should be processed successfully', () => {
  cy.get('.confirmation-message').should('contain', 'Payment successful');
});

Then('the payment should be declined', () => {
  cy.get('.error-message').should('contain', 'Invalid card details');
});

Then('the user should be redirected back to the confirmation page', () => {
  cy.url().should('include', '/confirmation');
});

Then('the user should be redirected back to the checkout page', () => {
  cy.url().should('include', '/checkout');
});

Then('the payment should not be processed', () => {
  cy.get('.error-message').should('contain', 'Payment not processed');
});

File: step_definitions/Multiple Payment Methods.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class PaymentSteps extends CheckoutPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is on the checkout page")
    public void theUserIsOnTheCheckoutPage() {
        navigateToCheckoutPage();
    }

    @When("the user selects credit card as the payment method")
    public void theUserSelectsCreditCardAsThePaymentMethod() {
        selectPaymentMethod("Credit Card");
    }

    @When("enters valid credit card details")
    public void entersValidCreditCardDetails() {
        enterCreditCardDetails(testHarness.getData("PaymentData", "ValidCardNumber"), testHarness.getData("PaymentData", "ExpiryDate"), testHarness.getData("PaymentData", "CVV"));
    }

    @Then("the payment should be processed successfully")
    public void thePaymentShouldBeProcessedSuccessfully() {
        verifyPaymentProcessed();
    }

    @Then("the user should receive a payment confirmation")
    public void theUserShouldReceiveAPaymentConfirmation() {
        verifyPaymentConfirmation();
    }

    @When("the user selects debit card as the payment method")
    public void theUserSelectsDebitCardAsThePaymentMethod() {
        selectPaymentMethod("Debit Card");
    }

    @When("enters valid debit card details")
    public void entersValidDebitCardDetails() {
        enterDebitCardDetails(testHarness.getData("PaymentData", "ValidDebitCardNumber"), testHarness.getData("PaymentData", "ExpiryDate"), testHarness.getData("PaymentData", "CVV"));
    }

    @When("the user selects PayPal as the payment method")
    public void theUserSelectsPayPalAsThePaymentMethod() {
        selectPaymentMethod("PayPal");
    }

    @When("logs into their PayPal account")
    public void logsIntoTheirPayPalAccount() {
        loginToPayPal(testHarness.getData("PayPalData", "Username"), testHarness.getData("PayPalData", "Password"));
    }

    @When("enters invalid credit card details")
    public void entersInvalidCreditCardDetails() {
        enterCreditCardDetails(testHarness.getData("PaymentData", "InvalidCardNumber"), testHarness.getData("PaymentData", "ExpiryDate"), testHarness.getData("PaymentData", "CVV"));
    }

    @Then("the payment should be declined")
    public void thePaymentShouldBeDeclined() {
        verifyPaymentDeclined();
    }

    @Then("the user should receive an error message")
    public void theUserShouldReceiveAnErrorMessage() {
        verifyErrorMessage();
    }

    @When("enters invalid debit card details")
    public void entersInvalidDebitCardDetails() {
        enterDebitCardDetails(testHarness.getData("PaymentData", "InvalidDebitCardNumber"), testHarness.getData("PaymentData", "ExpiryDate"), testHarness.getData("PaymentData", "CVV"));
    }

    @When("fails to log into their PayPal account")
    public void failsToLogIntoTheirPayPalAccount() {
        loginToPayPal(testHarness.getData("PayPalData", "InvalidUsername"), testHarness.getData("PayPalData", "InvalidPassword"));
    }

    @Then("the payment should not be processed")
    public void thePaymentShouldNotBeProcessed() {
        verifyPaymentNotProcessed();
    }
}